apps/meteor/app/apps/server/converters/threads.ts

```diff
@@ -0,0 +1,203 @@
+
+		const room = (await this.orch.getConverters().get('rooms').convertById(mainMessage.rid)) as IRoom | undefined;
+
+		if (!room) {
+			return [];
+		}
+
+		const convertToApp = cachedFunction(this.orch.getConverters().get('users').convertToApp.bind(this.orch.getConverters().get('users')));
+
+		const convertUserById = cachedFunction(this.orch.getConverters().get('users').convertById.bind(this.orch.getConverters().get('users')));
+
+		return Promise.all([mainMessage, ...replies].map((msg) => this.convertMessage(msg, room, convertUserById, convertToApp)));
+	}
+
+	async convertMessage(
+		msgObj: IMessage,
+		room: IRoom,
+		convertUserById: ReturnType<Orchestrator['users']>['convertById'],
+		convertToApp: ReturnType<Orchestrator['users']>['convertToApp'],
+	): Promise<AppsEngineMessage> {
+		const map = {
+			id: '_id',
+			threadId: 'tmid',
+			reactions: 'reactions',
+			parseUrls: 'parseUrls',
+			text: 'msg',
+			createdAt: 'ts',
+			updatedAt: '_updatedAt',
+			editedAt: 'editedAt',
+			emoji: 'emoji',
+			avatarUrl: 'avatar',
+			alias: 'alias',
+			file: 'file',
+			customFields: 'customFields',
+			groupable: 'groupable',
+			token: 'token',
+			blocks: 'blocks',
+			room,
+			editor: async (message: IMessage) => {
+				if (!isEditedMessage(message)) {
+					return undefined;
+				}
+
+				const { editedBy } = message;
+
+				return convertUserById(editedBy._id);
+			},
+			attachments: async (message: IMessage) => {
+				if (!message.attachments) {
+					return undefined;
+				}
+				const result = await this._convertAttachmentsToApp(message.attachments);
+				delete message.attachments;
+				return result;
+			},
+			sender: async (message: IMessage) => {
+				if (!message.u?._id) {
+					return undefined;
+				}
+
+				let user = await convertUserById(message.u._id);
+
+				// When the sender of the message is a Guest (livechat) and not a user
+				if (!user) {
+					user = await convertToApp(message.u);
+				}
+
+				return user;
+			},
+		};
+
+		return (await transformMappedData(msgObj, map)) as unknown as AppsEngineMessage;
+	}
+
+	async _convertAttachmentsToApp(attachments: ExcludeUndefined<IMessage['attachments']>) {
+		if (typeof attachments === 'undefined' || !Array.isArray(attachments)) {
+			return undefined;
+		}
+
+		const map = {
+			collapsed: 'collapsed',
+			color: 'color',
+			text: 'text',
+			timestampLink: 'message_link',
+			thumbnailUrl: 'thumb_url',
+			imageDimensions: 'image_dimensions',
+			imagePreview: 'image_preview',
+			imageUrl: 'image_url',
+			imageType: 'image_type',
+			imageSize: 'image_size',
+			audioUrl: 'audio_url',
+			audioType: 'audio_type',
+			audioSize: 'audio_size',
+			videoUrl: 'video_url',
+			videoType: 'video_type',
+			videoSize: 'video_size',
+			fields: 'fields',
+			actionButtonsAlignment: 'button_alignment',
+			actions: 'actions',
+			type: 'type',
+			description: 'description',
+			author: (attachment: ExcludeUndefined<IMessage['attachments']>[number]) => {
```

Comment:
Is there a difference in using this type expression instead of the plain `MessageAttachment` type?

Reply:
Exactly the same, there is no difference. but since we were manipulating IMesssage types, I thought it would make more sense to continue with IMessage usage